File: restack.go
package restack

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"strings"
)

// Request is a request to process a rebase instruction list.
type Request struct {
	// Name of the git remote. If set, an opt-in section that pushes restacked
	// branches to this remote will also be generated.
	//
	// This field is optional.
	RemoteName string

	// Input and output instruction lists.
	From io.Reader
	To   io.Writer
}

// Restacker processes the rebase instruction list.
type Restacker interface {
	Restack(context.Context, *Request) error
}

// GitRestacker restacks instruction lists using the provided Git instance.
type GitRestacker struct {
	// Controls access to Git commands.
	//
	// This field is required.
	Git Git
}

var _ Restacker = (*GitRestacker)(nil)

// Restack process the provided instruction list.
func (r *GitRestacker) Restack(ctx context.Context, req *Request) error {
	src := req.From

	rebasingBranch, err := r.Git.RebaseHeadName(ctx)
	if err != nil {
		return err
	}

	branches, err := r.Git.ListBranches(ctx)
	if err != nil {
		return err
	}

	knownBranches := make(map[string][]Branch)
	for _, b := range branches {
		knownBranches[b.Hash] = append(knownBranches[b.Hash], b)
	}

	gr := gitRestack{
		RemoteName:     req.RemoteName,
		RebaseHeadName: rebasingBranch,
		KnownBranches:  knownBranches,
		To:             req.To,
	}

	scanner := bufio.NewScanner(src)
	for scanner.Scan() {
		gr.Process(scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return err
	}

	gr.Finish()
	return nil
}

type gitRestack struct {
	RemoteName     string
	RebaseHeadName string
	KnownBranches  map[string][]Branch
	To             io.Writer

	lastLineBranches []Branch
	updatedBranches  []string
	wrotePushSection bool
}

func (r *gitRestack) Process(line string) error {
	if len(line) == 0 {
		// Empty lines delineate sections.
		// Push out any pending "git branch -x" statements
		// before printing the next section.
		var err error
		if !r.updatePreviousBranches() {
			// updatePreviousBranches adds a newline
			// after the git branch statements
			// if there were any new entries.
			// So we don't need to add another.
			_, err = fmt.Fprintln(r.To, line)
		}
		return err
	}

	// If we see comments, write the push section first.
	if len(line) > 0 && line[0] == '#' {
		r.updatePreviousBranches()
		if err := r.writePushSection(false, true); err != nil {
			return err
		}
	}

	// (p[ick]|f[ixup]|s[quash]) hash ...
	parts := strings.SplitN(line, " ", 3)
	if len(parts) > 1 {
		switch parts[0] {
		case "f", "fixup", "s", "squash":
			// Do nothing.
		default:
			r.updatePreviousBranches()
		}
	}

	// Most lines go in as-is.
	if _, err := fmt.Fprintln(r.To, line); err != nil {
		return err
	}

	if len(parts) < 2 {
		return nil
	}

	switch parts[0] {
	case "p", "pick", "r", "reword", "e", "edit":
		r.lastLineBranches = r.KnownBranches[parts[1]]
	}

	return nil
}

func (r *gitRestack) Finish() {
	r.updatePreviousBranches()
	r.writePushSection(true, false)
}

// Adds "git branch -f" directives for recorded branches, if any.
// Reports whether entries were added.
func (r *gitRestack) updatePreviousBranches() (updated bool) {
	branches := r.lastLineBranches
	r.lastLineBranches = nil

	for _, b := range branches {
		if b.Name == r.RebaseHeadName {
			continue
		}

		fmt.Fprintf(r.To, "exec git branch -f %v\n", b.Name)
		r.updatedBranches = append(r.updatedBranches, b.Name)
		updated = true
	}

	// Add an empty line between branch sections.
	if updated {
		fmt.Fprintln(r.To)
	}
	return updated
}

const _pushSectionPrefix = "# Uncomment this section to push the changes.\n"

func (r *gitRestack) writePushSection(padBefore, padAfter bool) error {
	if r.wrotePushSection {
		return nil
	}
	r.wrotePushSection = true

	if len(r.updatedBranches) == 0 || len(r.RemoteName) == 0 {
		return nil
	}

	if padBefore {
		io.WriteString(r.To, "\n")
	}
	if _, err := io.WriteString(r.To, _pushSectionPrefix); err != nil {
		return err
	}

	for _, b := range r.updatedBranches {
		if _, err := fmt.Fprintf(r.To, "# exec git push -f %s %s\n", r.RemoteName, b); err != nil {
			return err
		}
	}

	if padAfter {
		io.WriteString(r.To, "\n")
	}

	return nil
}

File: src/restack.rs
//! Implements the core restacking logic.

use std::collections::{HashMap, LinkedList};
use std::io::{self, BufRead, Write};
use std::path;

use anyhow::{Context, Result};

use crate::git;

#[cfg(test)]
mod tests;

/// Configures a Restack operation and provides the ability to run it.
pub struct Config<'a, Git>
where
    Git: git::Git,
{
    /// Defines how to talk to Git during the restack.
    git: Git,

    /// Current working directory inside which restack is running.
    cwd: &'a path::Path,
}

impl<'a, Git> Config<'a, Git>
where
    Git: git::Git,
{
    /// Builds a new restack configuration operating inside the given directory
    /// using the provided git object.
    pub fn new(cwd: &'a path::Path, git: Git) -> Self {
        Self { git, cwd }
    }

    /// Reads a git rebase instruction list from `src`
    /// and writes a new instruction list to `dst`
    /// that updates branches found at intermediate commits to their new positions.
    ///
    /// If `remote_name` is specified, includes an opt-in section
    /// at the bottom of the instruction list that
    /// updates the remotes for all affected branches.
    pub fn restack<I: io::Read, O: io::Write>(
        &self,
        remote_name: Option<&str>,
        src: I,
        dst: O,
    ) -> Result<()> {
        let rebase_branch_name = self
            .git
            .rebase_head_name(self.cwd)
            .context("Could not determine rebase head name")?;
        let all_branches = self
            .git
            .list_branches(self.cwd)
            .context("Unable to list branches")?;

        let mut known_branches: HashMap<&str, LinkedList<&git::Branch>> = Default::default();
        all_branches.iter().for_each(|b| {
            known_branches.entry(&b.shorthash).or_default().push_back(b);
        });

        let src = io::BufReader::new(src);
        let mut restack = Restack {
            remote_name,
            rebase_branch_name: &rebase_branch_name,
            dst: io::BufWriter::new(dst),
            known_branches: &known_branches,
            last_line_branches: Vec::new(),
            updated_branches: Vec::new(),
            wrote_push: false,
        };

        for line in src.lines() {
            let line = line.context("Failed while reading input")?;
            restack.process(&line)?;
        }

        restack.update_previous_branches()?;
        restack.write_push_section(true, false)?;

        Ok(())
    }
}

/// Holds state for an ongoing restack operation.
struct Restack<'a, O: io::Write> {
    /// Name of the remote, if any.
    remote_name: Option<&'a str>,

    /// Name of the branch we're rebasing.
    rebase_branch_name: &'a str,

    /// Destination writer.
    dst: io::BufWriter<O>,

    /// Known branches, keyed by their short hashes.
    known_branches: &'a HashMap<&'a str, LinkedList<&'a git::Branch>>,

    last_line_branches: Vec<&'a git::Branch>,
    updated_branches: Vec<&'a git::Branch>,
    wrote_push: bool,
}

impl<'a, O: io::Write> Restack<'a, O> {
    pub fn process(&mut self, line: &str) -> Result<()> {
        if line.is_empty() {
            // Empty lines delineate sections.
            // Write pending "git branch -x" statements
            // before going on to the next section.
            if !self.update_previous_branches()? {
                // update_previous_branches adds a trailing newline
                // only if git branch statements were added.
                // So if it didn't do anything, re-add the empty line.
                self.write_line("")?;
            }
            return Ok(());
        }

        // Comments usually mark the end of instructions.
        // Flush optional "git push" statements.
        if line.get(0..1) == Some("#") {
            self.update_previous_branches()?;
            self.write_push_section(false, true)
                .context("Could not write 'git push' section")?;
        }

        // (p[ick]|f[ixup]|s[quash]) hash ...
        let mut parts = line.splitn(3, ' ');

        let cmd = parts.next();
        if let Some(cmd) = cmd {
            match cmd {
                "f" | "fixup" | "s" | "squash" => {}, // do nothing
                _ => {
                    self.update_previous_branches()?;
                },
            }
        }

        // Most lines go as-is.
        self.write_line(line)?;

        let Some(cmd) = cmd else { return Ok(()); };
        let hash = match cmd {
            "p" | "pick" | "r" | "reword" | "e" | "edit" => match parts.next() {
                Some(s) => s,
                None => return Ok(()),
            },
            _ => return Ok(()),
        };

        if let Some(branches) = self.known_branches.get(hash) {
            self.last_line_branches.extend(branches);
        }

        Ok(())
    }

    fn write_push_section(&mut self, pad_before: bool, pad_after: bool) -> Result<()> {
        if self.wrote_push {
            return Ok(());
        }
        self.wrote_push = true;

        if self.updated_branches.is_empty() {
            return Ok(());
        }

        let Some(remote_name) = self.remote_name else { return Ok(()); };

        if pad_before {
            writeln!(self.dst)?;
        }
        writeln!(self.dst, "# Uncomment this section to push the changes.")?;
        for br in &self.updated_branches {
            writeln!(self.dst, "# exec git push -f {} {}", remote_name, br.name)?;
        }
        if pad_after {
            writeln!(self.dst)?;
        }

        Ok(())
    }

    /// Adds "exec git branch -f" statements to the instruction list.
    /// Reports whether any statements were added.
    fn update_previous_branches(&mut self) -> Result<bool> {
        let mut updated = false;
        for b in self.last_line_branches.drain(0..) {
            if b.name.as_str() == self.rebase_branch_name {
                continue;
            }

            writeln!(self.dst, "exec git branch -f {}", b.name)?;
            self.updated_branches.push(b);
            updated = true;
        }

        if updated {
            writeln!(self.dst)?;
        }

        Ok(updated)
    }

    fn write_line(&mut self, line: &str) -> Result<()> {
        writeln!(self.dst, "{}", line).map_err(Into::into)
    }
}
